File: algorithms/229.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
    vector<int> majorityElement(vector<int>& nums)
    {
        int a = 0, cnta = 0, b = 1, cntb = 0;
        for ( unsigned int i = 0; i < nums.size(); ++i )
        {
            if ( a == nums[i] )
                cnta += 1;
            else if ( b == nums[i] )
                cntb += 1;
            else if ( cnta == 0 )
                a = nums[i], cnta = 1;
            else if ( cntb == 0 )
                b = nums[i], cntb = 1;
            else
                cnta -= 1, cntb -= 1;
        }
        cnta = cntb = 0;
        for ( unsigned int i = 0; i < nums.size(); ++i )
        {
            if ( a == nums[i] ) cnta += 1;
            if ( b == nums[i] ) cntb += 1;
        }
        vector<int> result;
        if ( cnta > (int)nums.size()/3 ) result.push_back(a);
        if ( cntb > (int)nums.size()/3 ) result.push_back(b);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/rust/229.rs
impl Solution {
    #[allow(dead_code)]
    pub fn majority_element(nums: Vec<i32>) -> Vec<i32> {
        let (mut num_1, mut cnt_1) = (0, 0);
        let (mut num_2, mut cnt_2) = (1, 0);
        for &num in nums.iter() {
            if num_1 == num {
                cnt_1 += 1;
            } else if num_2 == num {
                cnt_2 += 1;
            } else if cnt_1 == 0 {
                num_1 = num;
                cnt_1 = 1;
            } else if cnt_2 == 0 {
                num_2 = num;
                cnt_2 = 1;
            } else {
                cnt_1 -= 1;
                cnt_2 -= 1;
            }
        }
        let (mut cnt_1, mut cnt_2) = (0, 0);
        for &num in nums.iter() {
            if num == num_1 {
                cnt_1 += 1;
            } else if num == num_2 {
                cnt_2 += 1;
            }
        }
        let mut result = vec![];
        if cnt_1 > nums.len() / 3 {
            result.push(num_1);
        }
        if cnt_2 > nums.len() / 3 {
            result.push(num_2);
        }
        result
    }
}

struct Solution;

fn main() {
    println!("Hello, world!");
}
